drivers/rtc/rtc_sam.c

```diff
@@ -0,0 +1,681 @@
+
+#include <string.h>
+#include <soc.h>
+
+#define RTC_SAM_REG_GET_FIELD(value, field) \
+	((RTC_##field##_Msk & value) >> RTC_##field##_Pos)
+
+#define RTC_SAM_WPMR_REGISTER (*((uint32_t *)(((uint8_t *)(SUPC)) + 0xD4)))
+#define RTC_SAM_WPMR_DISABLE 0x52544300
+#define RTC_SAM_WPMR_ENABLE 0x52544301
+
+#define RTC_SAM_CALIBRATE_PPB_MAX (1950000)
+#define RTC_SAM_CALIBRATE_PPB_MIN (-1950000)
+#define RTC_SAM_CALIBRATE_PPB_QUANTA (1500)
+#define RTC_SAM_CALIBRATE_PPB_LOW_SCALE (30500)
+
+typedef void (*rtc_sam_irq_init_fn_ptr)(void);
+
+struct rtc_sam_config {
+	Rtc *regs;
+	uint16_t irq_num;
+	rtc_sam_irq_init_fn_ptr irq_init_fn_ptr;
+};
+
+struct rtc_sam_data {
+#ifdef CONFIG_RTC_ALARM
+	rtc_alarm_callback alarm_callback;
+	void *alarm_user_data;
+#endif /* CONFIG_RTC_ALARM */
+#ifdef CONFIG_RTC_UPDATE
+	rtc_update_callback update_callback;
+	void *update_user_data;
+#endif /* CONFIG_RTC_UPDATE */
+	struct k_spinlock lock;
+	struct k_sem cr_sec_evt_sem;
+	struct k_sem cr_upd_ack_sem;
+};
+
+static void rtc_sam_disable_wp(void)
+{
+	RTC_SAM_WPMR_REGISTER = RTC_SAM_WPMR_DISABLE;
+}
+
+static void rtc_sam_enable_wp(void)
+{
+	RTC_SAM_WPMR_REGISTER = RTC_SAM_WPMR_ENABLE;
+}
+
+static bool rtc_sam_validate_tm(const struct rtc_time *timeptr, uint32_t mask)
+{
+	if ((mask & RTC_ALARM_TIME_MASK_SECOND) &&
+	    (timeptr->tm_sec < 0 || timeptr->tm_sec > 59)) {
+		return false;
+	}
+
+	if ((mask & RTC_ALARM_TIME_MASK_MINUTE) &&
+	    (timeptr->tm_min < 0 || timeptr->tm_min > 59)) {
+		return false;
+	}
+
+	if ((mask & RTC_ALARM_TIME_MASK_HOUR) &&
+	    (timeptr->tm_hour < 0 || timeptr->tm_hour > 23)) {
+		return false;
+	}
+
+	if ((mask & RTC_ALARM_TIME_MASK_MONTH) &&
+	    (timeptr->tm_mon < 0 || timeptr->tm_mon > 11)) {
+		return false;
+	}
+
+	if ((mask & RTC_ALARM_TIME_MASK_MONTHDAY) &&
+	    (timeptr->tm_mday < 1 || timeptr->tm_mday > 31)) {
+		return false;
+	}
+
+	if ((mask & RTC_ALARM_TIME_MASK_YEAR) &&
+	    (timeptr->tm_year < 0 || timeptr->tm_year > 199)) {
+		return false;
+	}
+
+	return true;
+}
+
+static uint32_t rtc_sam_timr_from_tm(const struct rtc_time *timeptr)
+{
+	uint32_t timr;
+
+	timr = RTC_TIMR_SEC(bin2bcd(timeptr->tm_sec));
+	timr |= RTC_TIMR_MIN(bin2bcd(timeptr->tm_min));
+	timr |= RTC_TIMR_HOUR(bin2bcd(timeptr->tm_hour));
+
+	return timr;
+}
+
+static uint32_t rtc_sam_calr_from_tm(const struct rtc_time *timeptr)
+{
+	uint32_t calr;
+	uint8_t centuries;
+	uint8_t years;
+
+	calr = RTC_CALR_DATE(bin2bcd(timeptr->tm_mday));
+	calr |= RTC_CALR_MONTH(bin2bcd(timeptr->tm_mon + 1));
+	centuries = (uint8_t)((timeptr->tm_year / 100) + 19);
+	years = (uint8_t)((timeptr->tm_year % 100));
+	calr |= RTC_CALR_CENT(bin2bcd(centuries));
+	calr |= RTC_CALR_YEAR(bin2bcd(years));
+	calr |= RTC_CALR_DAY(bin2bcd(timeptr->tm_wday + 1));
+	return calr;
+}
+
+static int rtc_sam_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	struct rtc_sam_data *data = dev->data;
+	const struct rtc_sam_config *config = dev->config;
```

Comment:
Use this convention:

```diff
// Define regs
+Rtc *regs = config->regs;
...
// Use it for less indirection readings on the code
-	config->regs->RTC_IER = RTC_IER_ACKEN;
+	regs->RTC_IER = RTC_IER_ACKEN;
```

Then apply everywhere.

Reply:
applied everywhere